src/khepri_projection.erl

```diff
@@ -0,0 +1,282 @@
+  when is_boolean(ReadConcurrency) ->
+    [{read_concurrency, ReadConcurrency} | Acc];
+to_ets_options(write_concurrency, WriteConcurrency, Acc)
+  when is_boolean(WriteConcurrency) orelse WriteConcurrency =:= auto ->
+    [{write_concurrency, WriteConcurrency} | Acc];
+to_ets_options(compressed, true, Acc) ->
+    [compressed | Acc];
+to_ets_options(compressed, false, Acc) ->
+    Acc;
+to_ets_options(Key, Value, _Acc) ->
+    throw({unexpected_option, Key, Value}).
+
+-spec name(Projection) -> Name when
+      Projection :: projection(),
+      Name :: atom().
+%% @doc Returns the name of the given `Projection'.
+%% @end
+%% This is a utility that's necessary because the `khepri_projection'
+%% record is opaque.
+
+name(#?MODULE{name = Name}) ->
+    Name.
+
+-spec init(Projection) -> Ret when
+      Projection :: projection(),
+      Ret :: ok | {error, exists}.
+%% @hidden
+%% Initializes a projection. The current implementation creates an ETS
+%% table using the projection's `name/1' and {@link options()}.
+
+init(#?MODULE{name = Name, ets_options = EtsOptions}) ->
+    case ets:info(Name) of
+        undefined ->
+            _ = ets:new(Name, EtsOptions),
+            ok;
+        _Info ->
+            {error, exists}
+    end.
+
+-spec trigger(Projection, Path, OldProps, NewProps) -> Ret when
+      Projection :: projection(),
+      Path :: khepri_path:native_path(),
+      OldProps :: khepri:node_props(),
+      NewProps :: khepri:node_props(),
+      Ret :: ok.
+%% @hidden
+%% Applies the projection function against the entry from the tree to return
+%% a projected record. This projected record is then applied to the ETS table.
+
+trigger(
+  #?MODULE{name = Name, project_fun = ProjectFun},
+  Path, OldProps, NewProps) ->
+    Table = ets:whereis(Name),
+    try
+        case ProjectFun#standalone_fun.arity of
+            2 ->
+                trigger_simple_projection(
+                  Table, ProjectFun, Path, OldProps, NewProps);
+            4 ->
+                Args = [Table, Path, OldProps, NewProps],
+                khepri_fun:exec(ProjectFun, Args),
+                ok
+        end
+    catch
+        Class:Reason:Stacktrace ->
+            Msg = io_lib:format("Failed to trigger projection~n"
+                                "  Projection: ~s~n"
+                                "  Path:~n"
+                                "    ~p~n"
+                                "  Old props:~n"
+                                "    ~p~n"
+                                "  New props:~n"
+                                "    ~p~n"
+                                "  Crash:~n"
+                                "    ~ts",
+                                [Name, Path, OldProps, NewProps,
+                                khepri_utils:format_exception(
+                                  Class, Reason, Stacktrace, #{column => 4})]),
+            ?LOG_ERROR(Msg, []),
+            ok
+    end.
```

Comment:
I think the try/catch block should wrap the various calls to `khepri_fun:exec()` only. Otherwise it will mask crashes in Khepri itself.